### PR TITLE
Modularize nonmodifying algorithms

### PR DESCRIPTION
**Pull request by miscco**
This PR has two goals

1. it starts to modularize the available non-modifying algorithms and adds tests so that we can have confidence in their implementation
2. It also moves 3 additional algorithms (fill_n, fill and lexicographical_compare) into their own files. With that we are able to remove the `<algorithm>` include from `<array>` which should improve compile times as we do not need to pull in algorithm everywhere

### CODE REVIEW

File: libcudacxx/include/cuda/std/detail/libcxx/include/array [hunk 1]
```diff
@@ -107,6 +107,9 @@ template <size_t I, class T, size_t N> const T&& get(const array<T, N>&&) noexce
 #include <__config>
 #endif // __cuda_std__
 
+#include "__algorithm/equal.h"
```
Comment (miscco) [inline libcudacxx/include/cuda/std/detail/libcxx/include/array:110]: feels-good-man

File: libcudacxx/include/cuda/std/detail/libcxx/include/__algorithm/fill.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+template <class _ForwardIterator, class _Tp>
+inline _LIBCUDACXX_HIDE_FROM_ABI _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR_AFTER_CXX11
+void fill(_ForwardIterator __first, _ForwardIterator __last, const _Tp& __value_)
+{
+    _CUDA_VSTD::__fill(__first, __last, __value_, typename iterator_traits<_ForwardIterator>::iterator_category());
+}
+
+_LIBCUDACXX_END_NAMESPACE_STD
+
+#endif // _LIBCUDACXX___ALGORITHM_FILL_H
```
Comment (wmaxey) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__algorithm/fill.h:51]: Missing newlines. NVHPC hates this.

**Comment by wmaxey** [comment]
Is this with clang format applying?

File: libcudacxx/.upstream-tests/test/std/algorithms/alg.nonmodifying/alg.any_of/any_of.pass.cpp [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        assert(cuda::std::any_of(cpp17_input_iterator<const int*>(ia),
+                           cpp17_input_iterator<const int*>(ia), test1()) == false);
+    }
+
+#if TEST_STD_VER > 11
+    static_assert(test_constexpr(), "");
+#endif
+
+  return 0;
+}
```
Comment (miscco) [inline libcudacxx/.upstream-tests/test/std/algorithms/alg.nonmodifying/alg.any_of/any_of.pass.cpp:71]: Missing newline
```suggestion
}

```

File: libcudacxx/.upstream-tests/test/std/algorithms/alg.nonmodifying/alg.find/find.pass.cpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+int main(int, char**)
+{
+    test();
+
+#if TEST_STD_VER > 11
+    static_assert(test(), "");
+#endif
+
+  return 0;
+}
```
Comment (miscco) [inline libcudacxx/.upstream-tests/test/std/algorithms/alg.nonmodifying/alg.find/find.pass.cpp:80]: MIssing newline
```suggestion
}

```

File: libcudacxx/.upstream-tests/test/std/algorithms/alg.nonmodifying/alg.find/find_if.pass.cpp [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+int main(int, char**)
+{
+    test();
+
+#if TEST_STD_VER > 11
+    static_assert(test(), "");
+#endif
+
+  return 0;
+}
```
Comment (miscco) [inline libcudacxx/.upstream-tests/test/std/algorithms/alg.nonmodifying/alg.find/find_if.pass.cpp:80]: Missing newline
```suggestion
}

```

File: libcudacxx/.upstream-tests/test/std/algorithms/alg.nonmodifying/alg.find/find_if_not.pass.cpp [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+int main(int, char**)
+{
+    test();
+
+#if TEST_STD_VER > 11
+    static_assert(test(), "");
+#endif
+
+  return 0;
+}
```
Comment (miscco) [inline libcudacxx/.upstream-tests/test/std/algorithms/alg.nonmodifying/alg.find/find_if_not.pass.cpp:80]: MIssing newline
```suggestion
}

```

File: libcudacxx/.upstream-tests/test/support/test_iterators.h [hunk 7]
```diff
@@ -1343,4 +1344,30 @@ ProxyRange(R&&) -> ProxyRange<cuda::std::views::all_t<R&&>>;
 
 #endif // TEST_STD_VER > 14
 
+namespace types {
+template <class Ptr>
+using random_access_iterator_list =
+    type_list<Ptr,
+#if TEST_STD_VER >= 20
```
Comment (miscco) [inline libcudacxx/.upstream-tests/test/support/test_iterators.h:1351]: Hm I believe this should actually be >= C++17

**Review by wmaxey — APPROVED** [review]

**Comment by miscco** [comment]
Has been merged together with #627

**Pull request closed without merging**